src/dataflow/src/sink/kafka.rs

```diff
@@ -382,92 +447,123 @@ impl KafkaSinkState {
-                    current_state
-                } else {
-                    SendState::Shutdown
+    // TODO: bound the backoff
+    async fn retry_on_txn_error<'a, F, Fut, T>(&self, f: F) -> Result<(), bool>
+    where
+        F: Fn(KafkaTxProducer) -> Fut,
+        Fut: Future<Output = KafkaResult<T>>,
+    {
+        let shutdown = Cell::new(false);
+        let self_producer = self.producer.clone();
+        let result = Retry::default()
+            .clamp_backoff(Duration::from_secs(60 * 10))
+            // Yes this is bad but we had an infinite loop before so it's no worse. Fix when
+            // addressing error strategy holistically.
+            .max_tries(usize::MAX)
+            .retry_async(|_| async {
+                match f(self_producer.clone()).await {
+                    Ok(_) => Ok(()),
+                    Err(KafkaError::Transaction(e)) => {
+                        if e.txn_requires_abort() {
+                            info!("Error requiring abort in kafka sink: {:?}", e);
+                            let self_self_producer = self_producer.clone();
+                            let should_shutdown = Retry::default()
+                                .clamp_backoff(Duration::from_secs(60 * 10))
+                                // Yes this is bad but we had an infinite loop before so it's no
+                                // worse.  Fix when addressing error strategy holistically.
+                                .max_tries(usize::MAX)
+                                .retry_async(|_| async {
+                                    match self_self_producer.abort_transaction().await {
+                                        Ok(_) => Ok(false),
+                                        Err(KafkaError::Transaction(e)) if e.is_retriable() => {
+                                            Err(KafkaError::Transaction(e))
+                                        }
+                                        Err(_) => Ok(true),
+                                    }
+                                })
+                                .await
+                                .unwrap_or(true);
+                            shutdown.set(should_shutdown);
+                            Ok(())
+                        } else if e.is_retriable() {
+                            info!("Retriable error in kafka sink: {:?}", e);
+                            Err(KafkaError::Transaction(e))
+                        } else {
+                            shutdown.set(true);
+                            Ok(())
+                        }
+                    }
+                    Err(_) => {
+                        shutdown.set(true);
+                        Ok(())
+                    }
                 }
-            }
-            _ => SendState::Shutdown,
+            })
+            .await;
+
+        // Consdier a retriable error that's hit our max backoff to be fatal.
+        if result.is_err() || shutdown.get() {
```

Comment:
I actually do think that `usize::MAX` is reasonable because in order to keep retrying, the transaction calls need to keep returning errors that rdkafka has marked as retriable.  If we can't complete our transaction calls, we can't really be making progress in the sink otherwise so we're not unnecessarily stalling.

I'm happy to discuss this more when I'm thinking about errors (I'll probably write up a short doc?) -- but if we're not on the same page, what do you think about leaving the behavior the same as it is and discussing more later?

Errors is the next thing I plan to work on (once I get through outstanding PRs) so I'm not worried about it being dropped as a TODO